### PR TITLE
Sparse histograms for continuous test

### LINKED ISSUE #4044: Let mimir-continuous-test generate native histograms
Mimir continuous test should be able to generate histograms beside float series

Requirements
- generation of float and histogram series can be controlled separatelly
- possible to generate float and histogram series at the same time
- be able to generate counter/gauge integer histograms and counter/gauge float histograms (4 different kinds)
- in gauge histograms the sine wave pattern can be encoded in bucket counter
- in counter histograms the sine wave pattern can be encoded in the rate of increase of bucket counter

### PR DESCRIPTION
**Pull request by zenador**
#### What this PR does

Adds histograms to continuous test.

As per the spec in the issue, samples and histograms are controlled separately with `-tests.write-read-series-test.float-samples-enabled` and `-tests.write-read-series-test.histogram-samples-enabled` respectively.

#### Which issue(s) this PR fixes or relates to

Fixes #4044

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by codesome — COMMENTED** [review]
On a high level, without going into details, I see that it would be good to modify the testing in these ways
* Add more buckets than just one. I added a suggestion. Also to add negative buckets.
* Bring real gauge-ness for gauge histogram (right now it is still a counter histogram with the gauge hint). TBD on how we set the bucket values that can be verified via queries.
* Histogram queries to return histograms as well (hence the verification code needs to change). This tests the returning of histogram sample via API, and queries working with buckets (instead of just the sum part of histogram).

File: pkg/continuoustest/metrics.go [hunk 1]
```diff
@@ -10,45 +10,45 @@ import (
...
 
 func NewTestMetrics(testName string, reg prometheus.Registerer) *TestMetrics {
 	return &TestMetrics{
-		writesTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		writesTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
 			Name:        "mimir_continuous_test_writes_total",
 			Help:        "Total number of attempted write requests.",
 			ConstLabels: map[string]string{"test": testName},
-		}),
+		}, []string{"metric_name"}),
```
Comment (krajorama) [inline pkg/continuoustest/metrics.go:27]: As discussed , we should call the additional label "type".
```suggestion
		}, []string{"type"}),
```

Also in all cases bellow

File: pkg/continuoustest/write_read_series.go [hunk 2]
```diff
@@ -246,84 +306,94 @@ func (t *WriteReadSeriesTest) runRangeQueryAndVerifyResult(ctx context.Context,
...
-		t.metrics.queryResultChecksFailedTotal.Inc()
+		t.metrics.queryResultChecksFailedTotal.WithLabelValues(metricName).Inc()
 		level.Warn(logger).Log("msg", "Range query result check failed", "err", err)
 		return errors.Wrap(err, "range query result check failed")
 	}
 	return nil
 }
 
-func (t *WriteReadSeriesTest) runInstantQueryAndVerifyResult(ctx context.Context, ts time.Time, resultsCacheEnabled bool) error {
+func (t *WriteReadSeriesTest) runInstantQueryAndVerifyResult(ctx context.Context, ts time.Time, resultsCacheEnabled bool, metricName, metricSumQuery string, generateValue generateValueFunc, records *MetricHistory) error {
```
Comment (krajorama) [inline pkg/continuoustest/write_read_series.go:330]:
```suggestion
func (t *WriteReadSeriesTest) runInstantQueryAndVerifyResult(ctx context.Context, ts time.Time, resultsCacheEnabled bool, typeLabel, metricSumQuery string, generateValue generateValueFunc, records *MetricHistory) error {
```

File: pkg/continuoustest/write_read_series.go [hunk 3]
```diff
@@ -246,84 +306,94 @@ func (t *WriteReadSeriesTest) runRangeQueryAndVerifyResult(ctx context.Context,
...
+	if records.lastWrittenTimestamp.IsZero() {
 		return alignTimestampToInterval(now, writeInterval)
 	}
 
-	return t.lastWrittenTimestamp.Add(writeInterval)
+	return records.lastWrittenTimestamp.Add(writeInterval)
 }
 
-func (t *WriteReadSeriesTest) findPreviouslyWrittenTimeRange(ctx context.Context, now time.Time) (from, to time.Time) {
+func (t *WriteReadSeriesTest) findPreviouslyWrittenTimeRange(ctx context.Context, now time.Time, metricName string, querySum querySumFunc, generateValue generateValueFunc) (from, to time.Time) {
```
Comment (krajorama) [inline pkg/continuoustest/write_read_series.go:390]:
```suggestion
func (t *WriteReadSeriesTest) findPreviouslyWrittenTimeRange(ctx context.Context, now time.Time, typeLabel string, querySum querySumFunc, generateValue generateValueFunc) (from, to time.Time) {
```

File: pkg/continuoustest/metrics.go [hunk 4]
```diff
@@ -10,45 +10,45 @@ import (
...
 		writesFailedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
 			Name:        "mimir_continuous_test_writes_failed_total",
 			Help:        "Total number of failed write requests.",
 			ConstLabels: map[string]string{"test": testName},
-		}, []string{"status_code"}),
-		queriesTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		}, []string{"status_code", "metric_name"}),
```
Comment (krajorama) [inline pkg/continuoustest/metrics.go:32]:
```suggestion
		}, []string{"status_code", "type"}),
```

File: pkg/continuoustest/metrics.go [hunk 5]
```diff
@@ -10,45 +10,45 @@ import (
...
+		queriesTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
 			Name:        "mimir_continuous_test_queries_total",
 			Help:        "Total number of attempted query requests.",
 			ConstLabels: map[string]string{"test": testName},
-		}),
-		queriesFailedTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		}, []string{"metric_name"}),
```
Comment (krajorama) [inline pkg/continuoustest/metrics.go:37]:
```suggestion
		}, []string{"type"}),
```

File: pkg/continuoustest/metrics.go [hunk 6]
```diff
@@ -10,45 +10,45 @@ import (
...
+		queriesFailedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
 			Name:        "mimir_continuous_test_queries_failed_total",
 			Help:        "Total number of failed query requests.",
 			ConstLabels: map[string]string{"test": testName},
-		}),
-		queryResultChecksTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		}, []string{"metric_name"}),
```
Comment (krajorama) [inline pkg/continuoustest/metrics.go:42]:
```suggestion
		}, []string{"type"}),
```

File: pkg/continuoustest/metrics.go [hunk 7]
```diff
@@ -10,45 +10,45 @@ import (
...
+		queryResultChecksTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
 			Name:        "mimir_continuous_test_query_result_checks_total",
 			Help:        "Total number of query results checked for correctness.",
 			ConstLabels: map[string]string{"test": testName},
-		}),
-		queryResultChecksFailedTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		}, []string{"metric_name"}),
```
Comment (krajorama) [inline pkg/continuoustest/metrics.go:47]:
```suggestion
		}, []string{"type"}),
```

File: pkg/continuoustest/metrics.go [hunk 8]
```diff
@@ -10,45 +10,45 @@ import (
...
+		queryResultChecksFailedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
 			Name:        "mimir_continuous_test_query_result_checks_failed_total",
 			Help:        "Total number of query results failed when checking for correctness.",
 			ConstLabels: map[string]string{"test": testName},
-		}),
+		}, []string{"metric_name"}),
```
Comment (krajorama) [inline pkg/continuoustest/metrics.go:52]:
```suggestion
		}, []string{"type"}),
```

File: pkg/continuoustest/write_read_series.go [hunk 9]
```diff
@@ -105,48 +148,65 @@ func (t *WriteReadSeriesTest) Run(ctx context.Context, now time.Time) error {
...
 		errs.Add(err)
-		err = t.runInstantQueryAndVerifyResult(ctx, ts, false)
+		err = t.runInstantQueryAndVerifyResult(ctx, ts, false, metricName, queryMetric, generateValue, records)
 		errs.Add(err)
 	}
-	return errs.Err()
 }
 
-func (t *WriteReadSeriesTest) writeSamples(ctx context.Context, timestamp time.Time) error {
+func (t *WriteReadSeriesTest) writeSamples(ctx context.Context, metricName string, timestamp time.Time, series []prompb.TimeSeries, records *MetricHistory) error {
```
Comment (krajorama) [inline pkg/continuoustest/write_read_series.go:200]:
```suggestion
func (t *WriteReadSeriesTest) writeSamples(ctx context.Context, typeLabel string, timestamp time.Time, series []prompb.TimeSeries, records *MetricHistory) error {
```

File: pkg/continuoustest/write_read_series.go [hunk 10]
```diff
@@ -173,70 +233,70 @@ func (t *WriteReadSeriesTest) writeSamples(ctx context.Context, timestamp time.T
...
-	ranges = append(ranges, [2]time.Time{randMinTime, randTime(randMinTime, t.queryMaxTime)})
+	randMinTime := randTime(adjustedQueryMinTime, records.queryMaxTime)
+	ranges = append(ranges, [2]time.Time{randMinTime, randTime(randMinTime, records.queryMaxTime)})
 	instants = append(instants, randMinTime)
 
 	return ranges, instants, nil
 }
 
-func (t *WriteReadSeriesTest) runRangeQueryAndVerifyResult(ctx context.Context, start, end time.Time, resultsCacheEnabled bool) error {
+func (t *WriteReadSeriesTest) runRangeQueryAndVerifyResult(ctx context.Context, start, end time.Time, resultsCacheEnabled bool, metricName, metricSumQuery string, generateValue generateValueFunc, records *MetricHistory) error {
```
Comment (krajorama) [inline pkg/continuoustest/write_read_series.go:295]:
```suggestion
func (t *WriteReadSeriesTest) runRangeQueryAndVerifyResult(ctx context.Context, start, end time.Time, resultsCacheEnabled bool, typeLabel, metricSumQuery string, generateValue generateValueFunc, records *MetricHistory) error {
```

File: pkg/continuoustest/write_read_series.go [hunk 11]
```diff
@@ -105,48 +148,65 @@ func (t *WriteReadSeriesTest) Run(ctx context.Context, now time.Time) error {
...
+	if t.cfg.WithHistograms {
+		for i := 0; i < 4; i++ {
+			t.RunInner(ctx, now, writeLimiter, errs, histogramMetricNames[i], querySumHist, generateHistogramSeries[i], generateHistogramValue[i], &t.histMetrics[i])
+		}
+	}
+
+	return errs.Err()
+}
+
+func (t *WriteReadSeriesTest) RunInner(ctx context.Context, now time.Time, writeLimiter *rate.Limiter, errs *multierror.MultiError, metricName string, querySum querySumFunc, generateSeries generateSeriesFunc, generateValue generateValueFunc, records *MetricHistory) {
```
Comment (krajorama) [inline pkg/continuoustest/write_read_series.go:164]:
```suggestion
func (t *WriteReadSeriesTest) RunInner(ctx context.Context, now time.Time, writeLimiter *rate.Limiter, errs *multierror.MultiError, typeLabel string, querySum querySumFunc, generateSeries generateSeriesFunc, generateValue generateValueFunc, records *MetricHistory) {
```

File: pkg/continuoustest/util.go [hunk 12]
```diff
@@ -10,13 +10,89 @@ import (
...
 	maxComparisonDelta = 0.001
 )
 
+type generateHistogramFunc func(t time.Time) prompb.Histogram
+type generateSeriesFunc func(name string, t time.Time, numSeries int) []prompb.TimeSeries
+type generateValueFunc func(t time.Time) float64
+type generateSampleHistogramFunc func(t time.Time) *model.SampleHistogram
+
+var (
+	generateHistogram = []generateHistogramFunc{
```
Comment (krajorama) [inline pkg/continuoustest/util.go:30]: Seems like these arrays need to line up, so wouldn't it be nicer to define something like
```
type HistogramProfile {
  metricName string
  typeLabel      string
  generate generateHistogramFunc
  generateSeries generateSeriesFunc
  generateValue generateValueFunc
  generateSample generateSampleHistogramFunc
}
```

And fill it out for each of the 4 ? that way these are made sure to be in sync

**Review by krajorama — COMMENTED** [review]
Looking good, two more cosmetic comments on the flags

**Review by krajorama — APPROVED** [review]
LGTM, let's ship it

**Pull request merged**